virtcontainers/device/drivers/vfio.go

```diff
@@ -57,44 +60,71 @@ func (device *VFIODevice) Attach(devReceiver api.DeviceReceiver) error {
 
 	// Pass all devices in iommu group
 	for _, deviceFile := range deviceFiles {
-
 		//Get bdf of device eg 0000:00:1c.0
 		deviceBDF, err := getBDF(deviceFile.Name())
 		if err != nil {
 			return err
 		}
-
-		device.BDF = deviceBDF
-
-		randBytes, err := utils.GenerateRandomBytes(8)
-		if err != nil {
-			return err
-		}
-		device.DeviceInfo.ID = hex.EncodeToString(randBytes)
-
-		if err := devReceiver.HotplugAddDevice(device, config.DeviceVFIO); err != nil {
-			deviceLogger().WithError(err).Error("Failed to add device")
-			return err
+		vfio := &config.VFIODev{
+			ID:  utils.MakeNameID("vfio", device.DeviceInfo.ID, maxDevIDSize),
+			BDF: deviceBDF,
 		}
+		device.vfioDevs = append(device.vfioDevs, vfio)
+	}
 
-		deviceLogger().WithFields(logrus.Fields{
-			"device-group": device.DeviceInfo.HostPath,
-			"device-type":  "vfio-passthrough",
-		}).Info("Device group attached")
+	// hotplug a VFIO device is actually hotplugging a group of iommu devices
+	if err := devReceiver.HotplugAddDevice(device, config.DeviceVFIO); err != nil {
```

Comment:
Why is this preferred than the original hotplug one-by-one interface? The interface is different than other device hotplug interfaces so I assume there is a valid reason for it. Please document it a bit in the above comments.

Reply:
That's because `HotplugAddDevice` accepts parameter of "device", in VFIO scenario, it's actually not a single device but all devices in same iommu group, and I need to save all the devices inside `device`. 
That's why I  modified the interface, and I think the comments from above "// Pass all devices in iommu group" already explained this well.

Reply:
Yes, I understand that. I was asking why `devReceiver.HotplugAddDevice` must take a slice of VFIO devices rather than just one device (like the other hotplug handlers) and then we can do iteration in the loop here like the original code does. Because in the end, you still call `s.hypervisor.hotplugAddDevice` in a loop, just at a different place.

Reply:
That will look weird, you see the `HotplugAddDevice` takes a `device` parameter, if I do the iteration here, we need to save one vfioDev inside `device`, and in next iteration, we need to modify the device struct and replace the original vfio dev with another dev, that will be annoying.

Besides that, for unplug part, if we don't save slice(all vfio devices in same IOMMU group), we need to do another "scan" again to get all vfio devices, and then another for loop to unplug them one by one. In my opinion the "scan" part can be avoided by saving all vfio dev into a slice in `device`, which can make the unplug logic very simple.